Rewrite State Sync, from a giant state machine to proper async code.

robin-near opened the pull request:
This rewrites state sync. All functionality is expected to continue to work without any protocol or database changes.

See the top of state/mod.rs for an overview.

State sync status is now available on the debug page; an example:
![Screenshot from 2024-09-24 09-48-37](https://github.com/user-attachments/assets/41c17e27-564d-403b-93ac-852297b70577)

shreyan-gupta commented on chain/client-primitives/src/types.rs:
```diff
@@ -245,28 +246,21 @@ pub fn format_shard_sync_phase(
-                f,
-                "StateSyncStatus {{ sync_hash: {:?}, shard_sync: {:?} }}",
-                self.sync_hash,
-                format_shard_sync_phase_per_shard(&self.sync_status, false)
```
Can delete `format_shard_sync_phase_per_shard`

shreyan-gupta commented on chain/client/src/sync/state/shard.rs:
```diff
@@ -0,0 +1,268 @@
+    return_if_cancelled!(cancel);
+
+    // Finalize; this needs to be done by the Chain.
+    *status.lock().unwrap() = ShardSyncStatus::StateApplyFinalizing;
```
Nice! With this setup, it should ideally be possible for us to convert ShardSyncStatus to a string instead? And we can keep adding more and more status with more information?

robin-near replied:
I didn't want to touch that part because it was also used in a couple of other places. Maybe I can clean it up after.

shreyan-gupta commented on chain/client/src/sync/state/shard.rs:
```diff
@@ -0,0 +1,268 @@
+/// would be blocked by the computation, thereby not allowing computation of other
+/// futures driven by the same driver to proceed. This function respawns the future
+/// onto the FutureSpawner, so the driver of the returned future would not be blocked.
+fn respawn_for_parallelism<T: Send + 'static>(
```
Instead of this, is it not possible to have small enough compute intensive tasks so as to not block the driver? This looks a bit odd overall where we are effectively "transferring" the future to a different spawner.awkward

robin-near replied:
That's basically what it does; the thing is that tokio_stream::iter never spawns anything; it only awaits multiple futures on the same driver.

shreyan-gupta commented on chain/client/src/sync/state/mod.rs:
```diff
@@ -0,0 +1,388 @@
+    /// and then in `run` we process them.
+    header_validation_queue: UnboundedReceiver<StateHeaderValidationRequest>,
+    chain_finalization_queue: UnboundedReceiver<ChainFinalizationRequest>,
+    chain_finalization_sender: UnboundedSender<ChainFinalizationRequest>,
```
Why do we need to store the sender as part of StateSync?

robin-near replied:
This is removed; replaced with AsyncSender.

shreyan-gupta commented on chain/client/src/sync/state/mod.rs:
```diff
@@ -0,0 +1,388 @@
+                    }
+                    Err(TryRecvError::Empty) => entry.get().status(),
+                },
+                Entry::Vacant(entry) => {
```
Could we encapsulate this into a function like `self.start_state_sync_for_shard` for better readability?

robin-near replied:
There is already the run_state_sync_for_shard function and the stuff here really isn't any useful logic other than just boilerplate code.

shreyan-gupta commented on chain/client/src/sync/state/mod.rs:
```diff
@@ -0,0 +1,388 @@
+    chain_finalization_sender: UnboundedSender<ChainFinalizationRequest>,
+
+    /// There is one entry in this map for each shard that is being synced.
+    shard_syncs: HashMap<(CryptoHash, ShardId), StateSyncShardHandle>,
```
Wait, where and how are we adding new entries into this?

robin-near replied:
We do a self.shard_syncs.entry(key) at the top, and if it's vacant we insert something into the entry.

shreyan-gupta reviewed (commented):
Looks great!

shreyan-gupta commented on chain/client/src/sync/state/mod.rs:
```diff
@@ -0,0 +1,388 @@
+    }
+
+    /// Processes the requests that the state sync module needed the Chain for.
+    fn process_chain_requests(&mut self, chain: &mut Chain) {
```
Slightly wary about having this setup for sync stuff happening...
While this is objectively better for our use case here, we have a pattern where we send an actix message to client to handle all things that should be in sync in client and this sorta breaks that... I can't think of better ways or alternatives

robin-near replied:
Changed to AsyncSender.

saketh-are commented on chain/client/src/sync/state/network.rs:
```diff
@@ -0,0 +1,294 @@
+                            part_id: *part_id,
+                        },
+                    );
+                    let state_value = PendingPeerRequestValue { peer_id: None, sender };
```
In the handling for NetworkRequests::StateRequestPart we select a specific peer from which to request the part, so it should be possible to store that and verify that the response comes back from the expected peer. However, it might be a bit ugly to pass the selected peer id back here from the network side of things, and I expect to redo how the state headers work soon, so I am fine with just leaving this as-is for now.

robin-near replied:
Yeah.. this part is awkward indeed. I also don't know what to do about it right now.

saketh-are commented on chain/network/src/peer/peer_actor.rs:
```diff
@@ -1100,7 +1100,14 @@ impl PeerActor {
                     .map(|response| PeerMessage::VersionedStateResponse(*response.0)),
                 PeerMessage::VersionedStateResponse(info) => {
                     //TODO: Route to state sync actor.
```
seems outdated

saketh-are commented on chain/client/src/sync/state/mod.rs:
```diff
@@ -0,0 +1,388 @@
+        let downloader = Arc::new(StateSyncDownloader {
+            clock,
+            store: store.clone(),
+            preferred_source: peer_source,
```
At the moment we are in an ugly situation with state headers; a node needs to be tracking **all** shards to serve headers for _any_ shard, nodes no longer track all shards, and the strategy for trying to obtain headers from the network is to request them from direct peers of the node at random. It is a remnant of the times when every peer tracked every shard, and it needs to be rewritten entirely. 

Before this PR, if an external source was available we would just directly get the headers from it without any attempts to get it from the network. It looks like we are changing that now and will try the network first for headers. I think it should be OK; just giving a heads up that there is a behavior change hidden here.

robin-near replied:
You mean before the PR, even if we configure to download from peers, we would always fetch the header from external? Hmm. OK.

robin-near commented on chain/client/src/sync/state/mod.rs:
```diff
@@ -0,0 +1,388 @@
+        store: Store,
+        epoch_manager: Arc<dyn EpochManagerAdapter>,
+        runtime: Arc<dyn RuntimeAdapter>,
+        network_adapter: AsyncSender<PeerManagerMessageRequest, PeerManagerMessageResponse>,
```
Changed to PeerMananagerAdapter.

robin-near commented on chain/client/src/sync/state/downloader.rs:
```diff
@@ -0,0 +1,226 @@
+                    )
+                    .await?;
+                let state_root = header.chunk_prev_state_root();
+                if runtime_adapter.validate_state_part(
```
That is out of scope of this refactoring that is already large :)

robin-near commented on chain/client/src/client.rs:
```diff
@@ -2465,34 +2464,40 @@ impl Client {
-                            network_adapter,
+                            self.runtime_adapter.store().clone(),
+                            self.epoch_manager.clone(),
+                            self.runtime_adapter.clone(),
```
We can look into that after yeah.

robin-near commented on tools/debug-ui/src/entity_debug/keys.tsx:
```diff
@@ -71,7 +71,7 @@ export function parseEntityKey(keyType: EntityKeyType, input: string): EntityKey
             return null;
         case 'trie_path':
-            if (/^s\d+[.]v\d+$\/[0-9A-Za-z]{44}\/[0-9a-f]*$/.test(input)) {
+            if (/^s\d+[.]v\d+\/[0-9A-Za-z]{43,44}\/[0-9a-f]*$/.test(input)) {
```
This fixes some random bug in the UI

robin-near commented on chain/network/src/client.rs:
```diff
@@ -71,10 +71,17 @@ pub struct StateRequestPart {
 #[rtype(result = "()")]
-pub struct StateResponse(pub Box<StateResponseInfo>);
+pub struct StateResponseReceived {
+    pub peer_id: PeerId,
```
In one place it's good to check that the peer ID is what we expected, to prevent spam.

robin-near commented on chain/client/src/client_actor.rs:
```diff
@@ -2079,13 +2044,21 @@ impl ClientActorInner {
 
         if block_hash == sync_hash {
             // The first block of the new epoch.
+            if let Err(err) = self.client.chain.validate_block(&block) {
```
Ah yes, actually I was supposed to do this in a separate PR, but forgot to separate this out. But it's already here and Saketh looked at it so I'll just leave it here.

robin-near commented on chain/client/src/sync/state/mod.rs:
```diff
@@ -0,0 +1,388 @@
+/// headers and parts in parallel for the requested shards, but externally, all that it exposes
+/// is a single `run` method that should be called periodically, returning that we're either
+/// done or still in progress, while updating the externally visible status.
+pub struct StateSync {
```
It is. There's nothing that is incompatible. It's removed anyway.

robin-near commented:
So @saketh-are @shreyan-gupta I addressed some of the issues and I'm just going to just merge this PR as it is right now, because the longer it takes the more merge conflicts I get. Then I'll follow up to clean up some of the stuff mentioned in the review comments.

pugachAG reviewed (approved):
Approving to unblock merging assuming @saketh-are and @shreyan-gupta already reviewed this
